src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -21,127 +24,137 @@
-            .set("teachlist", teachlist)
-            .build();
-
-      //update the information in the original entity
-      datastore.put(userUpdatedEntity);
-      response.sendRedirect("index.html");
+    String button_type = request.getParameter("submission").toString();
+    if (button_type == "update") {
+      this.doPut(request, response);
       return;
     }
 
-    @Override
-    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException { // read
-        Query<Entity> query =
-           Query.newEntityQueryBuilder().setKind("users").setOrderBy(OrderBy.desc("name")).build();
-        QueryResults<Entity> results = datastore.run(query);  
-
-        List<User> users = new ArrayList<>();
-        while (results.hasNext()) {
-          Entity entity = results.next();
-
-          String u_name = entity.getString("name");
-          String u_email = entity.getString("email");
-          String u_wishlist = entity.getString("wishlist");
-          String u_teachlist = entity.getString("teachlist");
-
-          User u_user = new User(u_name, u_email, u_wishlist, u_teachlist);
-          users.add(u_user);
-        }       
-
-        Gson gson = new Gson();
-        String json = gson.toJson(users);
-        response.setContentType("application/json;");
-        response.setCharacterEncoding("UTF-8");
-        response.getWriter().println(json);
-        
-        //maintians url and servlet info but doesn't run javascript
-        /*request.setAttribute("userJSON", json);
-        RequestDispatcher dispatcher = getServletContext()
-          .getRequestDispatcher("/index.html");
-        dispatcher.forward(request, response);*/
+    // get form information
+    String imgURL = Jsoup.clean(request.getParameter("imgURL"), Whitelist.none());
+    String name = Jsoup.clean(request.getParameter("name"), Whitelist.none());
+    String email = Jsoup.clean(request.getParameter("email"), Whitelist.none());
+    String[] wishlist = Jsoup.clean(request.getParameter("wishlist"), Whitelist.none()).split(",");
+    String[] teachlist = Jsoup.clean(request.getParameter("teachlist"), Whitelist.none()).split(",");
+    String id = Jsoup.clean(request.getParameter("id"), Whitelist.none());
+
+    // add kind users to keyfactory
+    keyFactory.setKind("users");
+
+    // take the user elements and create a new entity object for it
+    FullEntity userEntity = Entity.newBuilder(keyFactory.newKey(id)).set("imgURL", imgURL).set("name", name)
+        .set("email", email)
+        .set("wishlist", Arrays.asList(wishlist).stream().map(StringValue::of).collect(Collectors.toList()))
+        .set("teachlist", Arrays.asList(teachlist).stream().map(StringValue::of).collect(Collectors.toList())).build();
+
+    // insert that new entity
+    try {
+      datastore.add(userEntity);
+    } catch (Exception DatastoreException) {
+      this.doPut(request, response);
+      return;
     }
```

Comment:
Do not call PUT if `add()` fails. It may have failed **because** the entity is not new. You can write this as:
```suggestion
    datastore.add(userEntity);
```
Or if you need to handle the exception:
```suggestion
    try {
      datastore.add(userEntity);
    } catch (Exception DatastoreException) {
      e.printStackTrace();
      return;
    }
```

Reply:
ok gotcha, I handled the exception just in case.

Reply:
ok gotcha, I'll go with the first option